staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamicinformer
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamiclister"
+	"k8s.io/client-go/tools/cache"
+	"sync"
+	"time"
+)
+
+// NewDynamicSharedInformerFactory constructs a new instance of dynamicSharedInformerFactory for all namespaces.
+func NewDynamicSharedInformerFactory(client dynamic.Interface, defaultResync time.Duration) DynamicSharedInformerFactory {
+	return &dynamicSharedInformerFactory{
+		client: client,
+		defaultResync:    defaultResync,
+		informers: map[string]DynamicInformer{},
+	}
+}
+
+type dynamicSharedInformerFactory struct {
+	client    dynamic.Interface
+	defaultResync    time.Duration
+
+	lock             sync.Mutex
+	informers map[string]DynamicInformer
+}
+
+var _ DynamicSharedInformerFactory = &dynamicSharedInformerFactory{}
+
+func (f *dynamicSharedInformerFactory)  ForResource(gvr schema.GroupVersionResource, namespace string) DynamicInformer {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	informer, exists := f.informers[gvr.String()]
+	if exists {
+		return informer
+	}
+
+	// TODO(p0lyn0mial): should we accept cache.Indexer as an argument when constructing the factory ?!
```

Comment:
if we don't accept the indexer while constructing the object then I think the caller is unable to get object via `lister.Get()` (note: not `lister.Namespace().Get()`) because the object will be indexed by the namespace.

Reply:
> if we don't accept the indexer while constructing the object then I think the caller is unable to get object via `lister.Get()` (note: not `lister.Namespace().Get()`) because the object will be indexed by the namespace.

Wouldn't the namespace always be empty so the index is wasted (all under one key), but the keys themselves should match the lister's get index?

We don't want to change the signature on this method.

Reply:
I'll double check it.

Reply:
you were right, it works, sorry for the mix up.